cmd/juju/application/refresh.go

```diff
@@ -353,16 +363,27 @@ func (c *refreshCommand) Run(ctx *cmd.Context) error {
 	defer func() { _ = apiRoot.Close() }()
 
 	charmRefreshClient := c.NewCharmRefreshClient(apiRoot)
-	oldURL, oldOrigin, err := charmRefreshClient.GetCharmURLOrigin(ctx, c.ApplicationName)
-	if err != nil {
-		return errors.Trace(err)
-	}
 
 	// There is a timing window where deploy has been called and the charm
 	// is not yet downloaded. Check here to verify the origin has an ID,
-	// otherwise refresh result may be in accurate. We could use the
-	// retry package, but the issue is only seen in automated test due to
-	// speed. Can always use retry if it becomes an issue.
+	// otherwise refresh result may be in accurate.
+	// It should only occur in automated test, but if the default retries isn't enough,
+	// the returned error should help to investigate.
+	// Outside automated test, a new refresh will end up to works
+	var (
+		oldURL    *charm.URL
+		oldOrigin commoncharm.Origin
+	)
+	for range c.RetryGetCharmCount + 1 {
+		oldURL, oldOrigin, err = charmRefreshClient.GetCharmURLOrigin(ctx, c.ApplicationName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if oldOrigin.Source != commoncharm.OriginCharmHub || oldOrigin.ID != "" {
+			break
+		}
+		time.Sleep(c.RetryGetCharmDelay)
+	}
```

Comment:
We use https://github.com/juju/retry instead of a custom for loop.

Reply:
fixed